### PR TITLE
Fix network socket defaults

### LINKED ISSUE #4496: scsynth's network socket should bind to 127.0.0.1 by default rather than 0.0.0.0
## Environment

* SuperCollider version: 3.10.2

## Description

supercollider is intended to be run on local machines, and security is not really a top priority for the platform. however, right now, scsynth binds its UDP or TCP socket to `0.0.0.0` by default:

https://github.com/supercollider/supercollider/blob/develop/server/scsynth/scsynth_main.cpp#L140

which means that it listens on all interfaces. usually this means that anyone in your local network can mess with scsynth, which isn't what 99% of users want. even laptop ensembles tend to have everyone run their own client/server and most of the communication happens between clients.

simply changing this to `127.0.0.1` will be a more reasonable default configuration. this will not affect most users, but will impact people doing networked clients, who will need to set `-B 0.0.0.0`. see #4497.

disclaimer: i am not really great with network programming, so any insight from someone who knows better would be greatly appreciated!

## Securing an existing supercollider installation

until this is fixed, this hack can secure scsynth by placing it in your startup file:

`Server.program = "scsynth -B 127.0.0.1";`

## What about supernova?

supernova, as far as i can tell, does not support the `-B` flag, so getting supernova up to speed with scsynth is more involved. i'm not personally interested in maintaining supernova anymore, so i can't help with such issues.

### PR DESCRIPTION
**Pull request by snappizz**
## Purpose and Motivation

fix #4496 and #4497. see those issues for more detailed explanations. basic summary:

- scsynth and supernova listen to all network interfaces by default, which isn't good for security -- it means that by default, everyone on your LAN can crash your server! the defaults should only listen to your own machine
- scsynth supports the `-B` command line option, which should default to 127.0.0.1 and not 0.0.0.0
- supernova doesn't support `-B` at all, and in fact throws an error if you try to give it `-B`. (for this PR i only did enough to prevent the error, not fix the network issue in supernova)
- however `ServerOptions` doesn't expose `-B`, so i added `ServerOptions:bindAddress`

i was on the fence about whether to put this into 3.10.3 or 3.11, since it is a (minor) breaking change and a new feature. however it's also a pretty embarrassing security hole.

## Types of changes

- Documentation
- Bug fix
- New feature
- Breaking change

## To-do list

- [x] Code is tested
- [x] All tests are passing
- [x] Updated documentation
- [x] This PR is ready for review

### CODE REVIEW

**Comment by telephon** [comment]
Good! 

Please make sure the change is announced loudly enough, because it'll break most of the network music ensembles' setups. One of the complications that may come up especially in workshop and teaching situations is that different members will have different supercollider versions (especially different versions of 3.10) which are set up differently.

**Comment by telephon** [comment]
Also this breaks supernova for those setups, which may be a problem for some.

**Comment by snappizz** [comment]
> Please make sure the change is announced loudly enough, because it'll break most of the network music ensembles' setups.

yeah for sure

> Also this breaks supernova for those setups, which may be a problem for some.

in what way is supernova broken?

File: server/scsynth/scsynth_main.cpp [hunk 1]
```diff
@@ -51,7 +51,9 @@ void Usage() {
              "   -v print the supercollider version and exit\n"
              "   -u <udp-port-number>    a port number 0-65535\n"
              "   -t <tcp-port-number>    a port number 0-65535\n"
-             "   -B <bind-to-address>    an IP address\n"
+             "   -B <bind-to-address>\n"
+             "          Bind the UDP or TCP socket to this address.\n"
+             "          Default 127.0.0.1. Set to 0.0.0.0 to listen on all interfaces.\n"
```
Comment (lnihlen) [inline server/scsynth/scsynth_main.cpp]: Does it make sense to put doublequotes around these addresses to indicate they are string values?

File: server/supernova/server/server_args.cpp [hunk 2]
```diff
@@ -75,6 +75,7 @@ server_arguments::server_arguments(int argc, char* argv[]) {
 #endif
         ("restricted-path,P", value<vector<string> >(&restrict_paths), "if specified, prevents file-accessing OSC commands from accessing files outside <restricted-path>")
         ("threads,T", value<uint16_t>(&threads)->default_value(boost::thread::physical_concurrency()), "number of audio threads")
+        ("socket-address,B", value<string>()->default_value("127.0.0.1"), "reserved (not used)")
```
Comment (lnihlen) [inline server/supernova/server/server_args.cpp]: maybe a short note here indicating that supernova always listens to all incoming IPs?

**Comment by lnihlen** [comment]
So I patched this locally and was still able to get a remote computer to play sounds on my local scsynth instance. On the server:

```supercollider
(
~so = ServerOptions.new;
~so.maxLogins = 2;
s = Server.new('sec', options: ~so);
s.boot;
)
```

On the client:

```supercollider
(
s = Server.remote("roci", NetAddr("<ip addr of remote machine>", 57110));
{ SinOsc.ar }.play(s);
)
```

Plays even after the patch. Doing some further debugging.

**Comment by lnihlen** [comment]
> So I patched this locally and was still able to get a remote computer to play sounds on my local scsynth instance. On the server:
> 
> ```supercollider
> (
> ~so = ServerOptions.new;
> ~so.maxLogins = 2;
> s = Server.new('sec', options: ~so);
> s.boot;
> )
> ```
> 
> On the client:
> 
> ```supercollider
> (
> s = Server.remote("roci", NetAddr("<ip addr of remote machine>", 57110));
> { SinOsc.ar }.play(s);
> )
> ```
> 
> Plays even after the patch. Doing some further debugging.

Nope. Pilot error. Kindly disregard. This PR does what it says on the tin. 
Before the PR I can remotely play on the server with default options. After I cannot. Tested on Linux and MacOS.

**Comment by snappizz** [comment]
thanks @lnihlen!

i've decided that `socketAddress`, despite being a technically accurate name for this, deviates too much from the internal name of `bindTo` and the `-B` command-line flag, so i've compromised with `bindAddress`

ready 2 merge

**Pull request merged**